pkg/nydussdk/client.go

```diff
@@ -15,189 +16,140 @@ import (
+type NydusdClient interface {
+	GetDaemonInfo() (*types.DaemonInfo, error)
+
+	Mount(sharedMountPoint, bootstrap, daemonConfig string) error
 	Umount(sharedMountPoint string) error
-	GetFsMetric(sharedDaemon bool, sid string) (*model.FsMetric, error)
-}
 
-type NydusClient struct {
-	httpClient *http.Client
-}
+	BindBlob(daemonConfig string) error
+	UnbindBlob(daemonConfig string) error
 
-func NewNydusClient(sock string) (Interface, error) {
-	transport, err := buildTransport(sock)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to build transport for nydus client")
-	}
-	return &NydusClient{
-		httpClient: &http.Client{
-			Timeout:   defaultHTTPClientTimeout,
-			Transport: transport,
-		},
-	}, nil
-}
+	GetFsMetric(sharedDaemon bool, sid string) (*types.FsMetric, error)
 
-func (c *NydusClient) CheckStatus() (*model.DaemonInfo, error) {
-	addr := fmt.Sprintf("http://unix%s", infoEndpoint)
-	resp, err := c.httpClient.Get(addr)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to do HTTP GET from %s", addr)
-	}
-	defer resp.Body.Close()
-	b, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to read status response")
-	}
-	var info model.DaemonInfo
-	if err = json.Unmarshal(b, &info); err != nil {
-		return nil, err
-	}
-	return &info, nil
+	TakeOver() error
+	SendFd() error
+	Start() error
 }
 
-func (c *NydusClient) Umount(sharedMountPoint string) error {
-	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, sharedMountPoint)
-	req, err := http.NewRequest(http.MethodDelete, requestURL, nil)
-	if err != nil {
-		return errors.Wrap(err, "failed to create umount request")
-	}
-	resp, err := c.httpClient.Do(req)
-	if err != nil {
-		return errors.Wrapf(err, "failed to do HTTP DELETE to %s", requestURL)
-	}
-	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
-		return nil
-	}
-	return handleMountError(resp)
+// Nydusd API server http client used to command nydusd's action and
+// query nydusd working status.
+type nydusdClient struct {
+	httpClient *http.Client
 }
 
-func (c *NydusClient) GetFsMetric(sharedDaemon bool, sid string) (*model.FsMetric, error) {
-	var getStatURL string
+type query = url.Values
 
-	if sharedDaemon {
-		getStatURL = fmt.Sprintf("http://unix%s?id=/%s/fs", metricEndpoint, sid)
-	} else {
-		getStatURL = fmt.Sprintf("http://unix%s", metricEndpoint)
-	}
+func (c *nydusdClient) url(path string, query query) (url string) {
+	url = fmt.Sprintf("http://unix%s", path)
 
-	req, err := http.NewRequest(http.MethodGet, getStatURL, nil)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to create GetFsMetric request")
-	}
-	resp, err := c.httpClient.Do(req)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to do HTTP GET to %s", getStatURL)
-	}
-	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
-		return nil, errors.New("got unexpected http status StatusNoContent")
+	if len(query) != 0 {
```

Comment:
Better to use https://pkg.go.dev/net/url#example-URL (but it's not a blocker for the PR).